src/coprocessor/gadgets.rs

```diff
@@ -124,22 +134,364 @@ pub fn construct_list<F: LurkField, CS: ConstraintSystem<F>>(
+            not_dummy,
+            list.hash(),
+            &hash,
+        );
+        // recover the correct order
+        a_ptrs.reverse();
+    }
+    Ok(a_ptrs)
+}
+
+/// Deconstructs an `AllocatedPtr` with `car_cdr` semantics
+pub fn car_cdr<F: LurkField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    g: &GlobalAllocator<F>,
+    store: &Store<F>,
+    not_dummy: &Boolean,
+    maybe_cons: &AllocatedPtr<F>,
+) -> Result<(AllocatedPtr<F>, AllocatedPtr<F>), SynthesisError> {
+    let (car, cdr) = if not_dummy.get_value() == Some(true) {
+        let (car, cdr) = store
+            .car_cdr(&get_ptr(maybe_cons, store)?)
+            .expect("invalid Ptr");
+        (store.hash_ptr(&car), store.hash_ptr(&cdr))
+    } else {
+        (ZPtr::dummy(), ZPtr::dummy())
+    };
+
+    let nil = g
+        .get_allocated_ptr_from_ptr(&store.intern_nil(), store)
+        .expect("nil pointer not allocated");
+
+    let empty_str = g
+        .get_allocated_ptr_from_ptr(&store.intern_string(""), store)
+        .expect("empty string pointer not allocated");
+
+    let car = AllocatedPtr::alloc(&mut cs.namespace(|| "car"), || Ok(car))?;
+    let cdr = AllocatedPtr::alloc(&mut cs.namespace(|| "cdr"), || Ok(cdr))?;
+
+    let maybe_cons_is_nil =
+        maybe_cons.alloc_equal(&mut cs.namespace(|| "maybe_cons is nil"), &nil)?;
+
+    let maybe_cons_is_empty_str =
+        maybe_cons.alloc_equal(&mut cs.namespace(|| "maybe_cons is empty str"), &empty_str)?;
+
+    {
+        // when `maybe_cons` is nil, we enforce that `car` and `cdr` are both `nil`
+        let not_dumy_and_maybe_cons_is_nil = and(
+            &mut cs.namespace(|| "not_dummy and maybe_cons is nil"),
+            not_dummy,
+            &maybe_cons_is_nil,
+        )?;
+
+        car.implies_ptr_equal(
+            &mut cs.namespace(|| "maybe_cons is nil implies car is nil"),
+            &not_dumy_and_maybe_cons_is_nil,
+            &nil,
+        );
+        cdr.implies_ptr_equal(
+            &mut cs.namespace(|| "maybe_cons is nil implies cdr is nil"),
+            &not_dumy_and_maybe_cons_is_nil,
+            &nil,
+        );
+    }
+
+    {
+        // when `maybe_cons` is the empty string, we enforce that `car` is `nil`
+        // and `cdr` is the empty string
+        let not_dumy_and_maybe_cons_is_empty_str = and(
+            &mut cs.namespace(|| "not_dummy and maybe_cons is empty str"),
+            not_dummy,
+            &maybe_cons_is_empty_str,
+        )?;
+
+        car.implies_ptr_equal(
+            &mut cs.namespace(|| "maybe_cons is empty str implies car is nil"),
+            &not_dumy_and_maybe_cons_is_empty_str,
+            &nil,
+        );
+        cdr.implies_ptr_equal(
+            &mut cs.namespace(|| "maybe_cons is empty str implies cdr is empty str"),
+            &not_dumy_and_maybe_cons_is_empty_str,
+            &empty_str,
+        );
+    }
+
+    {
+        // when `maybe_cons` is not nil nor the empty string, we enforce the hash equality
```

Comment:
gratuitous English-language nit, for the sake of information: 'neither nil nor'

Reply:
You could also say 'not nil or'. Since you're using this construct in variable names below, I do think it should be corrected to something grammatical to avoid confusion.